### PR TITLE
Added unit tests and E2E tests

### LINKED ISSUE #378: Deleting a Bucket object with a missing bucket fails
/kind bug

**What steps did you take and what happened:**

_A clear and concise description of what the bug is and how has this been tested. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration_

1. Create a BucketClaim
2. Write an object to the bucket
3. Attempt to delete the Bucket which fails with an `BucketNotEmpty` error.
4. Delete the bucket in PC
5. Deletion fails with `NoSuchBucket`

**What did you expect to happen:**

I think the deletion should succeed if the bucket is missing, especially if in order to delete the bucket it needs to be empty.

**Anything else you would like to add:**

_Miscellaneous information that will assist in solving the issue._

Error from the logs:
```
E1223 16:15:04.824681       1 s3-handlers.go:105] "failed to delete bucket" err=<
        NoSuchBucket: The specified bucket does not exist.
                status code: 404, request id: 7b98496e-ed5d-4204-79c3-56f13e7517c3+210000+5349+523930+75312, host id: 7b98496e-ed5d-4204-79c3-56f13e7517c3+210000+5349+523930+75312

```

**Environment:**

- Nutanix objects version: 
- Nutanix COSI version:
  `main` built on 12/20/2024
- Kubernetes version: (use `kubectl version`):

### PR DESCRIPTION
**Pull request by ankush-ntx**
### Summary:
This PR introduces a testing framework for the Nutanix COSI driver, including both unit tests and end-to-end (E2E) tests. In addition, several minor bugs discovered during test development have been addressed and resolved.

### Changes:
**Unit Tests**:
- Covers core driver logic, IAM client ops, S3 client ops and provisioner methods.
- Mocks for S3 client and IAM client to isolate functionality.

**E2E Tests**:
- Tests all the workflows completely with edge cases.
- Including: Create/Delete buckets, Grant/Revoke user access
- Tests can be run on both a real cluster or a local cluster using Objects Triton as S3 backend objects store.

**CI/CD Pipeline:**
- Updated GitHub pipeline to include running unit tests when a PR is created.

**Deployment Changes:**
- `PC_SECRET` now only contains username and password credentials for the Prism Central in the previous format.
- `PC_ENDPOINT` instead should have the full endpoint of the Prism Central.

### Fixes:
- Fixed #378: Bucket deletion succeeds when bucket is missing.
- Fixed [CVE-2025-22868](https://github.com/advisories/GHSA-6v2p-p543-phr9)
- Added a few more relevant error statements.
- Minor fixes in bucket policy methods.

### How to run tests?
**Unit Tests**:
Execute the following to run unit tests:
```sh
go test ./...
```
To generate the coverage report and an HTML page to view the report:
```sh
go test -coverprofile=coverage.out ./...
go tool cover -html=coverage.out -o coverage.html
```
Open the HTML file in any web browser to view the coverage of each file.

**E2E Tests:**
Execute the following to run E2E tests:
```sh
sh scripts/setup_test_env.sh [flags]
```
```
Options:
-o, --oss_endpoint ENDPOINT    Nutanix Object Store instance endpoint, eg. "http://10.51.142.82:80".
-i, --pc_endpoint ENDPOINT     Prism Central endpoint, eg. "https://10.51.142.82:9440".
-u, --pc_user USERNAME         Prism Central username. [default = admin]
-p, --pc_pass PASSWORD         Prism Central password.
-a, --access_key KEY           Admin IAM Access key to be used for Nutanix Objects.
-s, --secret_key KEY           Admin IAM Secret key to be used for Nutanix Objects.
-n, --namespace NAMESAPCE      Cluster namespace for the COSI deployment [default = cosi]
```
Check the README.md for more info.

### CODE REVIEW

File: pkg/admin/user.go [hunk 1]
```diff
@@ -144,19 +144,26 @@ func (api *API) RemoveUser(ctx context.Context, uuid string) error {
...
+	if err != nil {
+		return fmt.Errorf("%w", err)
+	}
+
 	// Check response status
-	if delete_resp.StatusCode != 204 {
-		return fmt.Errorf("%s", delete_resp.Status)
+	if delete_resp.StatusCode == 404 {
+		return nil
+	} else if delete_resp.StatusCode != 204 {
```
Comment (anirudhbansal95) [inline pkg/admin/user.go]: nit: else if can be removed from line 165 since we are returning in line 164

File: pkg/admin/user_test.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+		resp, err := api.CreateUser(ctx, mockUsername, mockDisplayName)
+		assert.Error(t, err)
+		assert.Equal(t, admin.NutanixUserResp{}, resp)
+		assert.Contains(t, err.Error(), "user not created")
+	})
+
+	t.Run("TestCreateUser_MissingUsername", func(t *testing.T) {
+		api := &admin.API{}
+		_, err := api.CreateUser(ctx, "", mockDisplayName)
```
Comment (anirudhbansal95) [inline pkg/admin/user_test.go:94]: Currently we allow the display name to be empty, though not the user name. Is this the expected behaviour?

File: pkg/admin/util_test.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+		assert.Contains(t, err.Error(), "error while resolving endpoint")
+	})
+}
+
+func TestGetAccountName(t *testing.T) {
+	t.Run("TestGetAccountName_Success", func(t *testing.T) {
+		api := admin.API{
+			AccountName: "test",
+		}
+		accountName := api.AccountName
```
Comment (anirudhbansal95) [inline pkg/admin/util_test.go]: Maybe use api.GetAccountName() here and in the test below

File: pkg/driver/provisioner.go [hunk 4]
```diff
@@ -137,10 +140,12 @@ func (s *ProvisionerServer) DriverRevokeBucketAccess(ctx context.Context,
 
 	klog.InfoS("Deleting user", "id", req.GetAccountId())
 
-	err := s.ntnxIamClient.RemoveUser(ctx, req.GetAccountId())
+	err := s.NtnxIamClient.RemoveUser(ctx, req.GetAccountId())
 	if err != nil {
 		klog.ErrorS(err, "failed to delete user")
 	}
+
+	klog.InfoS("Successfully revoked access of user", "userName", req.GetAccountId(), "bucketName", req.GetBucketId())
```
Comment (anirudhbansal95) [inline pkg/driver/provisioner.go]: Do we also need to update the bucket policy here to remove the user from it, like we updated the bucket policy to grant user access in the previous case?

File: pkg/driver/provisioner.go [hunk 5]
```diff
@@ -137,10 +140,12 @@ func (s *ProvisionerServer) DriverRevokeBucketAccess(ctx context.Context,
 
 	klog.InfoS("Deleting user", "id", req.GetAccountId())
 
-	err := s.ntnxIamClient.RemoveUser(ctx, req.GetAccountId())
+	err := s.NtnxIamClient.RemoveUser(ctx, req.GetAccountId())
```
Comment (anirudhbansal95) [inline pkg/driver/provisioner.go]: Does ignoring the error mean this is only a best-effort API to revoke user access from bucket, since it returns nil error even if there was an IAM issue for example? Is the caller fine with this behaviour?

File: pkg/util/s3client/policy_test.go [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+		require.Len(t, modified.Statement, 3)
+
+		var foundSID1, foundSID2, foundSID3 bool
+		for _, ps := range modified.Statement {
+			switch ps.Sid {
+			case "sid1":
+				foundSID1 = true
+				assert.Equal(t, "Deny", string(ps.Effect))
+			case "sid2":
+				foundSID2 = true
```
Comment (anirudhbansal95) [inline pkg/util/s3client/policy_test.go:216]: Maybe also check that sid2 and sid3 have "Allow" effect

File: pkg/util/s3client/policy_test.go [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+		require.Len(t, ps.Resource, 1)
+		assert.Equal(t, "arn:aws:s3:::mybucket/*", ps.Resource[0])
+	})
+
+	t.Run("TestPolicyStatementMethods_AllowsEffect", func(t *testing.T) {
+		ps := s3client.NewPolicyStatement()
+		ps.Allows()
+		assert.Equal(t, s3client.Effect("Allow"), ps.Effect)
+		ps.Effect = "Other"
+		ps.Allows()
```
Comment (anirudhbansal95) [inline pkg/util/s3client/policy_test.go]: nit: Maybe add a comment that ps.Allows() doesn't modify an existing effect, if set.

File: pkg/util/s3client/s3-handlers_test.go [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+	t.Run("GetObjectInBucket_NotFound", func(t *testing.T) {
+		mock := &mocks.MockS3Client{
+			GetObjectFunc: func(input *s3.GetObjectInput) (*s3.GetObjectOutput, error) {
+				return nil, errors.New("not found")
+			},
+		}
+		s := &s3client.S3Agent{Client: mock}
+		content, err := s.GetObjectInBucket("missing-bucket", "missing-key")
+		assert.Error(t, err)
+		assert.Equal(t, "ERROR_ OBJECT NOT FOUND", content)
```
Comment (anirudhbansal95) [inline pkg/util/s3client/s3-handlers_test.go]: Is this a specific format in which the object not found error should be returned? Or can it be something like: ERROR_OBJECT_NOT_FOUND or some other string?

**Review by anirudhbansal95 — COMMENTED** [review]
Thanks Ankush. Reviewed the unit tests and the associated code. Yet to review the end to end tests.

File: tests/e2e/helpers/helpers.go [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+func CheckBucketStatusReady(ctx context.Context, bucketClient *bucketclientset.Clientset, bucketName string) error {
+	err := retry(ctx, func() error {
+		var err error
+
+		bucket, err := bucketClient.ObjectstorageV1alpha1().Buckets().Get(ctx, bucketName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		if bucket.Status.BucketReady == false {
```
Comment (anirudhbansal95) [inline tests/e2e/helpers/helpers.go]: nit: if !bucket.Status.BucketReady

File: tests/e2e/helpers/helpers.go [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+func CheckBucketStatusNotReady(ctx context.Context, bucketClient *bucketclientset.Clientset, bucketName string) error {
+	err := retry(ctx, func() error {
+		var err error
+
+		bucket, err := bucketClient.ObjectstorageV1alpha1().Buckets().Get(ctx, bucketName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		if bucket.Status.BucketReady == true {
```
Comment (anirudhbansal95) [inline tests/e2e/helpers/helpers.go]: nit: if bucket.Status.BucketReady

File: tests/e2e/revoke_suite_test.go [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+			// By("Checking if Bucket references the 'delete-bucketclass' and 'delete-bucketclaim'")
+			Expect(failBucket.Spec.BucketClassName).To(Equal(failBucketClass.Name))
+			Expect(failBucket.Spec.BucketClaim.Name).To(Equal(failBucketClaim.Name))
+
+			// By("Checking if Bucket status is 'bucketReady'")
+			err = helpers.CheckBucketStatusReady(ctx, bucketClient, failBucket.Name)
+			Expect(err).ToNot(HaveOccurred())
+
+			// By("Checking if Bucket is created in the Objectstore backend")
+			err = s3Client.WaitUntilBucketExists(&s3.HeadBucketInput{Bucket: &failBucket.Name})
```
Comment (anirudhbansal95) [inline tests/e2e/revoke_suite_test.go]: Do we need to call this function in other tests as well?

**Review by anirudhbansal95 — COMMENTED** [review]
Thanks Ankush. Took a look at the end to end tests. They look good in general with a few comments.

File: pkg/admin/user.go (hunk 1 above)
Comment (ankush-ntx) [reply pkg/admin/user.go]: Done.

File: pkg/admin/user_test.go (hunk 2 above)
Comment (ankush-ntx) [reply pkg/admin/user_test.go:94]: Yes. The API call is valid even with empty displayName.

File: pkg/admin/util_test.go (hunk 3 above)
Comment (ankush-ntx) [reply pkg/admin/util_test.go]: Done.

File: pkg/driver/provisioner.go (hunk 4 above)
Comment (ankush-ntx) [reply pkg/driver/provisioner.go]: Done. Changed policy SID to userId instead of username to facilitate policy removal when deleting bucket access.

File: pkg/driver/provisioner.go (hunk 5 above)
Comment (ankush-ntx) [reply pkg/driver/provisioner.go]: Fixed. Returning the error to the caller when failed to delete user.

File: pkg/util/s3client/policy_test.go (hunk 6 above)
Comment (ankush-ntx) [reply pkg/util/s3client/policy_test.go:216]: Done.

File: pkg/util/s3client/policy_test.go (hunk 7 above)
Comment (ankush-ntx) [reply pkg/util/s3client/policy_test.go]: Done.

File: pkg/util/s3client/s3-handlers_test.go (hunk 8 above)
Comment (ankush-ntx) [reply pkg/util/s3client/s3-handlers_test.go]: Fixed. Returning empty string now with error.

File: tests/e2e/helpers/helpers.go (hunk 9 above)
Comment (ankush-ntx) [reply tests/e2e/helpers/helpers.go]: Done.

File: tests/e2e/helpers/helpers.go (hunk 10 above)
Comment (ankush-ntx) [reply tests/e2e/helpers/helpers.go]: Done.

File: tests/e2e/revoke_suite_test.go (hunk 11 above)
Comment (ankush-ntx) [reply tests/e2e/revoke_suite_test.go]: CheckBucketExistenceInObjectstore()already waits for bucket creating, so removed WaitUntilBucketExists() as it is redundant.

File: pkg/admin/user.go [hunk 12]
```diff
@@ -90,27 +120,27 @@ func (api *API) CreateUser(ctx context.Context, username, display_name string) (
...
-		return result, fmt.Errorf("%s", resp.Status)
-	}
-
-	decodedResponse, err := ioutil.ReadAll(resp.Body)
+	decodedResponse, err := io.ReadAll(resp.Body)
 	if err != nil {
 		return result, fmt.Errorf("%w", err)
 	}
 
+	// Check respsonse status
```
Comment (anirudhbansal95) [inline pkg/admin/user.go]: nit: response

File: pkg/util/s3client/s3-handlers_test.go [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+`
+
+func TestNewS3Agent(t *testing.T) {
+	accessKey := "dummyAccessKey"
+	secretKey := "dummySecretKey"
+
+	t.Run("TestNewS3Agent_ValidSecureConnection", func(t *testing.T) {
+		// Insecure false and endpoint starting with "https" should succeed with a valid PEM CA certificate.
+		endpoint := "https://127.0.0.1:9440"
+		agent, err := s3client.NewS3Agent(accessKey, secretKey, endpoint, validPEMCert, false, false)
```
Comment (anirudhbansal95) [inline pkg/util/s3client/s3-handlers_test.go]: nit: Annotate the last two arguments of NewS3Agent

**Review by anirudhbansal95 — COMMENTED** [review]
Thanks Ankush. A few more comments. Yet to review the e2e tests.

**Review by anirudhbansal95 — APPROVED** [review]
Thanks Ankush, reviewed the e2e tests as well. Looks good to me with the comments above.

File: .github/workflows/unit-test.yaml [hunk 14]
```diff
@@ -0,0 +1,29 @@
+name: Unit Test
+
+on:
+  push:
+    branches: [ main ]
+  pull_request:
+    branches: [ main ]
```
Comment (manoj-nutanix) [inline .github/workflows/unit-test.yaml]:
```suggestion
  pull_request:
```
let's run on all PRs as unit test is not heavy job

File: pkg/util/s3client/s3-handlers.go [hunk 15]
```diff
@@ -126,9 +139,15 @@ func (s *S3Agent) DeleteBucket(name string) (bool, error) {
 		Bucket: aws.String(name),
 	})
 	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			klog.InfoS("DEBUG: after s3 call", "ok", ok, "aerr", aerr)
```
Comment (manoj-nutanix) [inline pkg/util/s3client/s3-handlers.go]:
```suggestion
```
Please remove debug logs which are not needed.

File: tests/e2e/helpers/helpers.go [hunk 16]
```diff
@@ -0,0 +1,392 @@
...
+		if i > 0 {
+			time.Sleep(sleep)
+		}
+		err = f()
+		if err == nil {
+			return nil
+		}
+	}
+	return err
+}
```
Comment (manoj-nutanix) [inline tests/e2e/helpers/helpers.go]: see if you can use https://pkg.go.dev/k8s.io/client-go/util/retry#OnError instead

File: tests/e2e/helpers/helpers.go [hunk 17]
```diff
@@ -0,0 +1,392 @@
...
+	err := retry(ctx, func() error {
+		var err error
+
+		_, err = http.Get(ossEndpoint)
+		if err != nil {
+			return err
+		}
+		_, err = s3Client.ListBuckets(&s3.ListBucketsInput{})
+		if err != nil {
+			return nil
```
Comment (manoj-nutanix) [inline tests/e2e/helpers/helpers.go]: should be?
```suggestion
			return err
```

if there's error in listing buckets, func `VerifyObjectstore` returns nil, which I believe is not intended

**Review by manoj-nutanix — APPROVED** [review]

File: tests/e2e/helpers/helpers.go (hunk 17 above)
Comment (ankush-ntx) [reply tests/e2e/helpers/helpers.go]: Done.

File: tests/e2e/helpers/helpers.go (hunk 16 above)
Comment (ankush-ntx) [reply tests/e2e/helpers/helpers.go]: Done.

File: pkg/util/s3client/s3-handlers.go (hunk 15 above)
Comment (ankush-ntx) [reply pkg/util/s3client/s3-handlers.go]: Done.

File: .github/workflows/unit-test.yaml (hunk 14 above)
Comment (ankush-ntx) [reply .github/workflows/unit-test.yaml]: Done.
